### ISSUE TITLE
Transpiler eliminates measurements and wrongly remaps output bitstrings

### ISSUE THREAD

**Issue by gabrieleagl** [Qiskit/qiskit-terra#8810]
### Environment

- **Qiskit Terra version**: 0.20.1
- **Python version**: 3.9.12
- **Operating system**: MacOS 12.6 (21G115)

### What is happening?

I transpile a circuit for a given provider, say FakeMumbai, with optimization level = 3.
The circuit has mid-measurements.

Before transpiling, it has 2 measurements, while after transpiling it has just 1 measurement.
This may be due to an approximation of a unitary: indeed, one of the measurement provides 1 almost every time.

The error consists in that after the removal of the measurement, output bitstrings are no longer consistent with those of the original circuit. In particular, the first circuit ouputs `{'10': 2, '00': 3, '11': 2556, '01': 2439}` while the second one `{'00': 2542, '01': 2458}`.

### How can we reproduce the issue?

Run the following code:

```
from qiskit import QuantumCircuit
qc = QuantumCircuit.from_qasm_file('qasm.txt')

from qiskit import transpile
from qiskit.providers.aer import AerSimulator
qasm_backend = AerSimulator()
qasm_circuit = transpile(qc, qasm_backend, optimization_level=3)
print(qasm_circuit.count_ops())
qasm_backend.run(qasm_circuit, shots=5000).result().get_counts()

from qiskit.providers.fake_provider import FakeMumbai 
fake_backend = FakeMumbai()
fake_circuit = transpile(qc, fake_backend, optimization_level=3)
print(fake_circuit.count_ops())

#Note that the transpiled circuit is run on qasm_backend again, no noise effect is in place:
qasm_backend.run(fake_circuit, shots=5000).result().get_counts()
```

Outputs are:

```
OrderedDict([('cx', 786), ('tdg', 333), ('t', 261), ('u2', 115), ('u3', 55), ('u', 54), ('h', 41), ('mcphase', 32), ('unitary', 27), ('ry', 12), ('p', 7), ('measure', 2), ('cp', 1), ('reset', 1)])
{'10': 2, '00': 3, '11': 2556, '01': 2439}
OrderedDict([('rz', 1312), ('cx', 1003), ('sx', 662), ('x', 60), ('measure', 1), ('reset', 1)])
{'01': 555, '00': 4445}
```

(Sorry for the big circuit attached, I tried to create a smaller one but couldn't replicate the issue)
[qasm.txt](https://github.com/Qiskit/qiskit-terra/files/9678009/qasm.txt)

### What should happen?

If '11' and '01' are the most frequent measurements in the first circuit, they should be also the outputs of the second circuit.

**Comment by jakelishman** [comment]
Thanks for the report! Could you try this with an updated Terra? The latest version is 0.21.2. It feels like this issue may be the one fixed by #8041, which was released in Terra 0.20.2.

**Comment by gabrieleagl** [comment]
Hi @jakelishman I confirm that the behavior is not replicated in the latest version. Thanks a lot!
